ports-of-call/math_utils.hpp

```diff
@@ -0,0 +1,94 @@
+//------------------------------------------------------------------------------
+// © 2021-2024. Triad National Security, LLC. All rights reserved.  This
+// program was produced under U.S. Government contract 89233218CNA000001
+// for Los Alamos National Laboratory (LANL), which is operated by Triad
+// National Security, LLC for the U.S.  Department of Energy/National
+// Nuclear Security Administration. All rights in the program are
+// reserved by Triad National Security, LLC, and the U.S. Department of
+// Energy/National Nuclear Security Administration. The Government is
+// granted for itself and others acting on its behalf a nonexclusive,
+// paid-up, irrevocable worldwide license in this material to reproduce,
+// prepare derivative works, distribute copies to the public, perform
+// publicly and display publicly, and to permit others to do so.
+//------------------------------------------------------------------------------
+
+#ifndef PORTS_OF_CALL_MATH_UTILS_HPP_
+#define PORTS_OF_CALL_MATH_UTILS_HPP_
+
+#include <cmath>
+#include <limits>
+#include <ports-of-call/portability.hpp>
+#include <ports-of-call/robust_utils.hpp>
+
+namespace PortsOfCall {
+namespace Math {
+
+// Faster implementation of std::pow for arithmetic bases and non-negative integer
+// exponents.  For sufficiently large integer powers std::pow may be faster, but testing
+// indicates that the following implementation is significantly faster (roughly a factor
+// of two or better) up to powers of at least 100.
+template <typename BaseT, typename ExponentT,
+          typename std::enable_if<std::is_arithmetic_v<std::decay_t<BaseT>> &&
+                                  std::is_integral_v<std::decay_t<ExponentT>>>::type * =
+              nullptr>
+PORTABLE_FORCEINLINE_FUNCTION constexpr auto power(BaseT base,
+                                                                     ExponentT exponent) {
+  using PowT = decltype(std::pow(base, exponent));
+  if (!Robust::check_nonnegative(exponent) || exponent > ExponentT{100}) {
```

Comment:
I think there may be an additional optimization for negative exponents where you do `base = 1 / base;` and `exponent = abs(exponent);` and not have to fall back on the slow/robust pow. I know this PR is merged so perhaps something to do in the future @Lona85 @Yurlungur

Reply:
Very good suggestion. Thanks @dholladay00 !